### ISSUE TITLE
[Feature request] Common stereo metrics.

### ISSUE THREAD

**Issue by copaah** [kornia/kornia#1123]
## 🚀 Feature
To make kornia attractive to stereo researcher, we should consider adding a set of commonly used stereo metrics as is seen in the [KITTI](http://www.cvlibs.net/datasets/kitti/eval_scene_flow.php?benchmark=stereo) or [Middlebury](https://vision.middlebury.edu/stereo/eval3/) benchmarks.

Metrics to consider to add:
* RootMeanSquaredError
* MeanAbsoluteError
* MeanBad{n}Error 

We should consider whether we want the metrics to inherit from [torchmetrics](https://torchmetrics.readthedocs.io/en/latest/l), such that the metrics can be used in `Lightning` modules.

## Motivation

Most new stereo research all implement their own metrics - this is a good opportunity for Kornia to get a foothold in stereo research.

## Pitch

In `kornia.utils.metrics` add above mentioned metrics. Discuss whether we want to inherit from `torchmetric`. The metric logic should ideally be split into functions and then referenced in a class that inherits from `torchmetric` in case we go that way. 

## Alternatives
N/A

## Additional context
N/A

**Comment by shijianjian** [comment]
Hi @copaah,

Thanks for the proposal. Those losses are common and not only for stereo images. Since we are a computer vision library, I personally think we should include losses that are more "CV" centric like SSIM, PSNR, etc. In that case, we may also differentiate from other libraries, convincing other packages to use us since there is no competition. Is there any stereo-specific losses/metrics we can include?

Jian

**Comment by edgarriba** [comment]
fully agree with @shijianjian - we should try to cover try cover "only" vision metrics/losses that interests our community.

Are the metrics you mention contained in `torchmetrics` ? Ideally we should find a tradeoff between libraries/implementations. /cc @Borda Any objection here ?

**Comment by copaah** [comment]
@edgarriba @shijianjian 

Thank you both for your inputs. I understand your objections.

The metrics I mentioned are not in `torchmetrics`. They would inherit from the `Metric` class like this:

```

class RootMeanSquaredError(torchmetrics.Metric):
    def __init__(self):
        self.add_state("root_sum_squared_error", default=torch.zeros(1))
        self.add_state("n_obs", default=torch.zeros(1))

    def update(self, preds, target):
        sum_squared_error = torch.pow(preds - target, 2)
        self.root_sum_squared_error += torch.sqrt(sum_squared_error.mean())
        self.n_obs += 1

    def compute(self):
        return torch.true_divide(self.root_sum_squared_error, self.n_obs)
```

And would be compatible with all `pytorch-lightning` functionality.

The metrics I mentioned are also part of `opencv-contrib`:
* [MeanBad{n}Error](https://docs.opencv.org/master/da/d17/group__ximgproc__filters.html#ga6db28b4258a9a820d5cf75b63cd51543)
* [MeanSquaredError](https://docs.opencv.org/master/da/d17/group__ximgproc__filters.html#gaad2210013ec0118e542cb38a01cae416)

I think these are the most stereo specific metrics that are commonly used and also evident from the KITTI and Middlebury benchmarks.

**Comment by shijianjian** [comment]
Hi @copaah,

Is the MeanBad{n}Error more close to be stereo-only? I think probably we can include some functions around computing disparity maps? @edgarriba what do you think?

Best,
Jian

**Comment by edgarriba** [comment]
I think it's not a bad idea - torchmetrics has only pytorch dependency and `Metric` seems a quite generic thing. We should re-think also about losses which is very close to metrics so that we can reuse the same functionality to backpropagate gradients.

Could we do somehow optional the `torchmetrics` import part ?

**Comment by copaah** [comment]
I would say `MeanBad{n}Error` is closer to stereo-only, yes.

I think a lot of this discussion stems from uncertainty about overall roadmap from `stereo` module in `kornia`. My initial hope was that `kornia` could be the go-to library for (deep learning based) stereo researcher and it would include all "helper" functionality such as metrics, the `StereoCamera` class and data loaders for common stereo datasets. This will unify stereo research going forward and will give `kornia` a good footprint in this line of computer vision.

Imagine stereo researchers could do something like this, when they research a new deep learning stereo model:

```
import pytorch_lightning as pl

from kornia.metrics.stereo import *
from kornia.stereo import KittiLoader, StereoCamera
from torchmetrics import MetricCollection

kitti_metrics = MetricCollection([
MeanBadNError(n=1),
MeanBadNError(n=2),
Density(),
MeanAbsError()
])   # Could even consider adding this to Kornia to make it even easier and standard across research

class CustomStereoModel(pl.LightningModule)
     def __init__(self):
            self.my_stereo_network = MyStereoNetwork()
            self.metrics = kitti_metrics

trainer  = pl.Trainer(...)
stereo_model = CustomStereoModel()
trainer.fit(stereo_model, train_dataloader=KittiLoader)

test_stereo_observation = KittiLoader[0]
stereo_camera = StereoCamera(test_stereo_observation.left_camera, test_stereo_observation.right_camera)
my_prediction = stereo_model(test_stereo_observation.left_image, test_stereo_observation.right_image)
point_cloud = stereo_camera.repoject_disparity_to_3D(my_prediction)
```

**Comment by shijianjian** [comment]
I would avoid using names like ```metrics``` and ```losses``` since they look like more DL related., probably go for ```kornia.measure.stereo```. BTW, what is the plan on the stereo? To have a ```kornia.stereo``` module? or only to include those metrics?

I like this decoupling that we do not import ```torchmetrics``` as all.

**Comment by edgarriba** [comment]
This sounds attractive - however I would like to be a bit conservative at the time to add stuff like datasets since we have been avoiding this from the beginning and just focus on curated methods. For datasets we could have an external/optional module. It's not the case here, but basically we would like to have quite a robust kornia core and not fall in the trap to have in the short term deprecated codes - research goes too fast and we do not aim to have every state of the art thing.

**Comment by copaah** [comment]
@shijianjian : What is the motivation behind avoiding DL nomenclature?

@edgarriba : Sounds good. I won't add any datasets - if I really want to, I could maybe do in a tutorial :)

How to proceed with this feature request? Should we just put in the backlog and wait for the `stereo` roadmap to be bit more mature? I don't see this as a high priority right now and would like to finish the `StereoCamera` class before adding other features.

**Comment by edgarriba** [comment]
yes - let's continue this discussion once we have more things in the stereo module

**Comment by Borda** [comment]
btw, we have started adding image metrics to TrochMetrics :]
https://github.com/PyTorchLightning/metrics

**Comment by copaah** [comment]
> btw, we have started adding image metrics to TrochMetrics :]
> https://github.com/PyTorchLightning/metrics

Do you think these metrics would be a welcome contribution in `torchmetrics`?

**Comment by shijianjian** [comment]
> @shijianjian : What is the motivation behind avoiding DL nomenclature?

Not a strong objection. The term of "loss" in DL is more or less ready to be called with ```loss.backward()```. If we are going for losses, we'd better have some support/tutorials/examples for that. Also, I think naming like ```kornia.measures``` would attract some DL libraries like ```torchmetrics``` to use us with less burden (that we are aiming on different directions).

**Comment by kristijanbartol** [comment]
I guess this feature is now implemented in https://github.com/PyTorchLightning/metrics/issues/799, except for MeanBad{n}Error metric? Should we also attempt to implement this third metric in torchmetrics instead of here?

**Comment by Borda** [comment]
> Should we also attempt to implement this third metric in torchmetrics instead of here?

that would be awesome :purple_heart:

**Comment by edgarriba** [comment]
if it fits in lightning world why but more stereo geometry stuff is coming soon to kornia so i'd say we will add it sooner or later

**Comment by Borda** [comment]
> if it fits in lightning world why but more stereo geometry stuff is coming soon to kornia so i'd say we will add it sooner or later

just to be clear, TM is coming from PL authors but it has zero dependencies to PL :)

**Comment by edgarriba** [comment]
our recent policy is just to have our (kornia non-profit org) libs dependencies except for fundamental tensor backend for now.

**Comment by Saharsh1005** [comment]
I would like to take this issue up, is it still available?

**Comment by edgarriba** [comment]
@Saharsh1005  feel free to take it

**Comment by Big-Marvel** [comment]
Hi @edgarriba @copaah 
Is this issue still open? If so, I would like to take this up.

**Comment by edgarriba** [comment]
@Big-Marvel what exact metric you want to implement?

**Comment by Big-Marvel** [comment]
@edgarriba I would like to implement MeanBad{n}Error as this is the most stereo-specific metric also this metric is widely used and standardized.
Let me know your thoughts on this.

**Comment by edgarriba** [comment]
Sounds good

**Comment by Big-Marvel** [comment]
Hi @edgarriba 
I have few clarification question for implementing MeanBad{n}Error
1. Should it be named mean_bad_error with a parameter n, or a more specific name like bad_pixel_error?
2. Is this specifically for stereo/disparity or general vector fields (like optical flow)? 
3. Should the metric handle an optional mask (like ignoring occluded pixels or zero-disparity regions)?
4. For MeanBad{n}, should the "mean" be computed per image in the batch or across the entire batch?

**Comment by edgarriba** [comment]
@Big-Marvel can you check if there are existing implementations, or how the datasets mentioned above use it ? this will help to defne a better api

**Comment by kirandharmar867-maker** [comment]
helo, if it is open i would like to sovle this one.
You just say "OK".
Thank You.

**Comment by edgarriba** [comment]
@kirandharmar867-maker if you are interested in contributing please provide concrete plans (and join discord). Your replies across the issues will make us suspect you are bot not following the code of conduct of the project

**Comment by kirandharmar867-maker** [comment]
> [@kirandharmar867-maker](https://github.com/kirandharmar867-maker) if you are interested in contributing please provide concrete plans (and join discord). Your replies across the issues will make us suspect you are bot not following the code of conduct of the project

sorry About by generic replies -- that was not my intention,
i am a genuine contributor and I understand the concern.
also i joined to discord.
thank You.

**Comment by shijianjian** [comment]
Hi @kirandharmar867-maker, are you working on this?

**Comment by VanniLeonardo** [comment]
Hi, I’d like to work on this.

I inspected the current `kornia.metrics` structure and I think the lowest-risk first PR for me would be a pure Kornia implementation without adding a hard `torchmetrics` dependency.

Proposed scope:
- add a new `kornia/metrics/disparity.py`
- add functional metrics:
  - `mean_absolute_disparity_error`
  - `root_mean_squared_disparity_error`
  - `mean_bad_pixel_error`
  - optionally `kitti_d1_error` for the KITTI 3px / 5% criterion
- support optional `valid_mask`
- support `reduction="mean" | "sum" | "none"`
- add tests under `tests/metrics/test_disparity.py`
- export from `kornia.metrics.__init__`
- add docs under the metrics documentation

I would avoid `torchmetrics.Metric` in the first PR to keep the dependency surface small. Maybe then a TorchMetrics-compatible wrapper could be discussed if you believe it's needed.

For the initial implementation I would follow the existing functional metric style, similar to `endpoint_error.py`.

A few design choices I would use unless you prefer otherwise:
- accept arbitrary same-shaped disparity tensors rather than forcing `(B, 1, H, W)`
- use boolean `valid_mask`, broadcastable to the disparity shape
- for masked `reduction="none"`, return the full error map with invalid positions zeroed
- return bad-pixel error as a fraction in `[0, 1]`, not a percentage
- keep `nn.Module` wrappers out of the first PR unless you want them included

**Comment by edgarriba** [comment]
@VanniLeonardo feel free to work on that

**Comment by VanniLeonardo** [comment]
Thanks @edgarriba . I have a first implementation ready locally following the scope above. Could you assign the issue to me before I open the PR, since the PR template asks for assignment?
